### PR TITLE
localstore: support seek reverse

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

File: kv/buffer_store.go [hunk 1]
```diff
@@ -63,12 +63,20 @@ func (s *BufferStore) Seek(k Key) (Iterator, error) {
...
+		return nil, errors.Trace(err)
+	}
+	retrieverIt, err := s.r.SeekReverse(k)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return newUnionIter(buferIt, retrieverIt, true), nil
 }
 
 // WalkBuffer iterates all buffered kv pairs.
```
Comment (shenli) [inline kv/buffer_store.go]: Need to add ReverseWalkBuffer?

File: kv/union_iter.go [hunk 2]
```diff
@@ -25,14 +25,16 @@ type UnionIter struct {
...
 }
 
-func newUnionIter(dirtyIt Iterator, snapshotIt Iterator) *UnionIter {
+func newUnionIter(dirtyIt Iterator, snapshotIt Iterator, reverse bool) *UnionIter {
 	it := &UnionIter{
 		dirtyIt:       dirtyIt,
 		snapshotIt:    snapshotIt,
 		dirtyValid:    dirtyIt.Valid(),
 		snapshotValid: snapshotIt.Valid(),
+		reverse:       reverse,
```
Comment (shenli) [inline kv/union_iter.go]: dirtyNext and snapshotNext should consider it.reverse?

File: kv/buffer_store.go (hunk 1 above)
Comment (coocood) [inline kv/buffer_store.go]: No, `WalkBuffer` just iterator all buffer data, the order doesn't matter.

File: kv/union_iter.go (hunk 2 above)
Comment (coocood) [inline kv/union_iter.go]: Iterators were created as reverse iterators.

File: store/localstore/boltdb/boltdb_test.go [hunk 3]
```diff
@@ -151,3 +151,42 @@ func (s *testSuite) TestSeek(c *C) {
...
+	c.Assert(v, BytesEquals, []byte("1"))
+
+	k, v, err = s.db.SeekReverse([]byte("bb"))
+	c.Assert(err, IsNil)
+	c.Assert(k, BytesEquals, []byte("b"))
+	c.Assert(v, BytesEquals, []byte("1"))
+
+	k, v, err = s.db.SeekReverse([]byte("b"))
+	c.Assert(err, NotNil)
+	c.Assert(string(k), Equals, "")
```
Comment (shenli) [inline store/localstore/boltdb/boltdb_test.go]: c.Assert(k, IsNil)
c.Assert(v, IsNil) ?

File: store/localstore/engine/engine.go [hunk 4]
```diff
@@ -39,6 +39,9 @@ type DB interface {
 	// Seek searches for the first key in the engine which is >= key in byte order, returns (nil, nil, ErrNotFound)
 	// if such key is not found.
 	Seek(key []byte) ([]byte, []byte, error)
+	// SeekReverse searches for the first key in the engine which is less than key in byte order, returns (nil, nil, ErrNotFound)
```
Comment (shenli) [inline store/localstore/engine/engine.go]: "SeekReverse searches the engine in backward order for the first key-value pair which key is less than the key in byte order, returns (nil, nil, ErrNotFound)" is better?

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@disksing mock-tikv should do the same thing. PTAL

File: store/localstore/kv.go [hunk 5]
```diff
@@ -58,14 +56,33 @@ func (s *dbStore) Seek(key []byte, startTS uint64) ([]byte, []byte, error) {
...
+// SeekReverse searches for the first key in the engine which is less than key in byte order.
+// Returns (nil, nil, ErrNotFound) if such key is not found.
+func (s *dbStore) SeekReverse(key []byte, startTS uint64) ([]byte, []byte, error) {
+	err := s.prepareSeek(startTS)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	s.wg.Done()
+	return s.db.SeekReverse(key)
+}
```
Comment (disksing) [inline store/localstore/kv.go]: I think you want to `SeekReverse()` before `wg.Done()`, like the `Seek()` method above.

**Comment by coocood** [comment]
@disksing PTAL

File: store/localstore/snapshot.go [hunk 6]
```diff
@@ -120,27 +151,40 @@ func (s *dbSnapshot) BatchGet(keys []kv.Key) (map[string][]byte, error) {
...
+	var (
+		k   kv.Key
+		v   []byte
+		err error
+	)
+	if reverse {
+		k, v, err = s.reverseMvccSeek(key)
+		if bytes.Equal([]byte(k), []byte(key)) {
+			log.Errorf("rev seek got equal key %x", k)
+		}
```
Comment (disksing) [inline store/localstore/snapshot.go]: Remove L182-184 ?

**Comment by disksing** [comment]
LGTM

**Comment by coocood** [comment]
@BusyJay PTAL

**Pull request merged**